### ISSUE TITLE
Interrogation about Chrome extension manifest v3 proposal meaning for Nano.

### ISSUE THREAD

**Issue by mikhoul** [NanoAdblocker/NanoCore#238]
Hi,

I really like Nano more than UBO but I was asking to myself what will be your direction if the proposal is adopted and that the actual API is neutered ?

My understanding is that Chrome (for me ) would become inferior to Firefox for adblockers (unless Firefox implement the same restriction which I doubt). 

If the change happen they can happen in the next 12 months which is relatively fast so the solutions/alternatives I've see are:

- Make an official version for Firefox and drop Chrome.
- Make an Edge version with their new browser with blink engine IF they don't have the same restriction.
- Submit to the new API and have an inefficient adblocker comparing to today.
- Forking Chromium keeping the old API as it is today and have a version especially build for Nano.

Maybe there is other solutions that I don't foresee. 🤔

What is your take on this @jspenguin2017 ?

**Comment by jspenguin2017** [comment]
I see 3 options: 
- Use the new API
- Fork Chromium
- Make a native app

I'll explain in detail when I have time.

**Comment by jspenguin2017** [comment]
Option 1: Use the new API

If the new API is really weak, then I don't think Firefox can survive for long. If websites can more easily get around the weak API of Chrome, then website owners will have a motivation to push everyone to use Chrome. And soon Chromium-based browsers will be the only usable browsers. 

If the new API is OK, then converting to the new API is probably going to be the easiest solution. 

The current proposal of new APIs are on the weak side, although they are not that bad. If the 30k limit is lifted and rules can be dynamically modified, then I would say it's OK.

**Comment by jspenguin2017** [comment]
Option 2: Fork Chromium

I would assume code for manifest v2 will eventually be removed, so the fork will not be a simple flag flip. Maintaining a browser fork is really difficult in general, and vulnerabilities can be opened if not careful. 

Another way is to use Electron to build a new browser. Unlike WebExtension, Electron provides really powerful APIs, which is unlikely to go away. However, the amount of work to maintain a browser is likely more than that of a fork, but since I don't have to use native code, vulnerabilities are less likely to occur.

**Comment by jspenguin2017** [comment]
Option 3: Make a native app

There's a shell there's a way. I'm sad to see that there's no open source native app that does more than DNS blocking. 

A native app is the best way to get around the constantly changing browser APIs, and as a bonus, it'll work for all browsers and even other apps. 

I'm not sure how hard this will be, last time I tried, it didn't go so well.

**Comment by jspenguin2017** [comment]
I'm probably will go with option 1 or 3, as for the timeline, I don't think it'll happen within 12 months. Right now, the APIs are not released yet, and the phase out period will likely take more than a year. 

I think it's a good time to get started on option 3, if things go wrong, option 1 can be used as fallback.

**Comment by mikhoul** [comment]
First thanks for your very detailed answer. :+1: 

The problem I see with option 3 if my understanding is fine is that a native app would not be able to remove some objects like cosmetic filters are doing ? 

For me a native app sound like when I use AdAway on android, I'm pretty sure I'm missing something here about how would work a native app  🤔  :question: 

I would miss "Dynamic filtering" since I use it occasionally to un-break some  filters that are good but that on specific website cause problems.

I don't think Firefox will disappear, it can lose users but Google will always keep Firefox alive as a caution to prevent a split of their business by US Gov (like Microsoft have done with Apple). 

IMO if Google weaken their API for blocking ads it could revigorante Firefox with power-users and gain 2-3% of users. When I migrated from Firefox to Chromium I dragged with me ~150 users and those users today recommend Chrome/chromium to their friends and I'm pretty sure I'm not alone.

But Mozilla management is crazy those day they removed so many things/features that I migrated to Chromium to stop from suffering from the many memory leaks in Firefox. So in one way it would not surprise me that they do some crazy thing in the future like following the very same Chrome API unless the management at Mozilla change.

For the fork I was thinking it was more easy than you said it is to keep the old API, like some module you glue to the main code when you build  chromium and that don't require lot of maintenance. I've seen [Slimjet](https://www.slimjet.com) a Chromium fork and with an integrated ad-blocker so I was thinking it was "easy" to integrate in a fork.

Time will tell. :octocat:

**Comment by jspenguin2017** [comment]
> native app would not be able to remove some objects like cosmetic filters are doing

Again, there's a shell there's a way. Native apps are (theoretically) able to do cosmetic filtering as they can patch everything the browser receives, however, I do agree that for cosmetic filtering, it's easier to implement using extensions. 

There is nothing stopping native apps from loading extensions into browsers, so if needed, the native app can have a companion extension. 

> sound like when I use AdAway on android

AdAway is a DNS manager, it is really limited in what it can do. I'm talking about native apps that can repack HTTPS requests. 

> For the fork I was thinking it was more easy than you said

I never tried. Maybe you're right, I don't know. 
If you have more experience with the Chromium code base maybe you can start a Chromium fork?

**Comment by jspenguin2017** [comment]
Actually instead of trying to reinventing the wheel and fail miserably, I'll probably just take a [wheel](https://github.com/mitmproxy/mitmproxy) off the shelf. 

I'll probably still need a JavaScript runtime though, as RegExp in filters are JS RegExp. For a prototype it'll be fine, if it works out I'll worry about optimization later.

**Comment by mikhoul** [comment]
Thanks for your complete and detailed answer it's enlighting and it remove interrogations I had.  👍

So it would be a little bit how AdGuard work, I never used AdGuard but my understanding is that there is a component that work at the OS level and another part as an extension for better efficiency.

> If you have more experience with the Chromium code base maybe you can start a Chromium fork?

I don't have experience but it is something I 'm thinking about for 2-3 months to be able to modify the Chromium GUI and keep it tight, maybe I will give it a try this year,  this issue will ad some motivation make this project higher in my priority.

Regards :octocat:

**Comment by onmyouji** [comment]
> Another way is to use Electron to build a new browser. Unlike WebExtension, Electron provides really powerful APIs, which are unlikely to go away. However, the amount of work to maintain a browser is likely more than that of a fork, but since Electron apps use JavaScript and not native code, vulnerabilities are less likely to occur.

I don't really know about Electron, but I remember [this talk](https://www.youtube.com/watch?v=Qirdy1TP1Rw) about building a web browser in Electron. The security engineer actually recommends not try to do it.

**Comment by nl255** [comment]
>Option 1: Use the new API
>If the new API is really weak, then I don't think Firefox can survive for long. If websites can more easily get around the weak APIs of Chrome, then website owners will have a motivation to push everyone to use Chrome. And soon Chromium-based browsers will probably become the only usable browsers.

Are you saying that websites will simply block Firefox users entirely if Firefox insists on keeping the existing webRequest API?


>Option 3: Make a native app
>There's a shell there's a way. I'm sad to see that there's no open source native app that does more than DNS blocking.
>A native app is the best way to get around the constantly changing browser APIs, and as a bonus, it'll probably work for all browsers and even non-browser apps.
>I'm not sure how hard this will be though, last time I tried it didn't go so well.

For this to work it would need to support HTTPS interception/rewriting which while it can be done opens up a huge can of worms security wise, especially if it is not done properly.  It would also require having a companion GUI app for creating and installing the custom SSL certificate used for said interception as most people aren't going to be able to do it on their own using the standard command line tools.  Even worse, on Android by default apps will not trust user installed root certificates (Chrome does but many other browsers and apps do not) and will need to be patched (unless the device is rooted, then the certificate can be installed to the system certificate store either directly or with a Magisk module).

**Comment by mikhoul** [comment]
@jspenguin2017   I have my answer for a fork on chromium to bypass a weak API if it was needed it'S possible and [Brave](https://github.com/brave/brave-browser) seem to already use this strategy:

A developer on the Brave team was explaining it on Reddit:

_______________________________________________________________________
_Bat-chriscat
BAT TEAM[M] 34 points 1 day ago* RemindMe!

It's worth noting that our **Brave Shields (ad blocker) is not an extension; it is natively implemented. So extension API changes leave our shields unaffected.**

Edit: We can always remove any code or update we don't like from the Chromium base we use. So even if this didn't just affect extensions but something deeper, we could just exclude it._
____________________________________________________________________________________

**Source**: https://old.reddit.com/r/brave_browser/comments/aijqm4/chrome_may_soon_change_how_3rd_party_ad_blockers/eeopqpb/


So maybe if they already do that you could fork Brave to integrate natively Nano inside and bypassing the API to intercept the calls you need ? 🤔  

This way it would be future proof and the control would be very granular to block everything. User would just install the browser and use it.  ☺️


Regards :octocat:

**Comment by nhantrn** [comment]
@jspenguin2017 You might want to checkout Proxydomo hosted here on github. https://github.com/amate/Proxydomo

It's a local proxy with https filtering capability. I've been using it to strip out <script /> among other filters of blacklisted pages before it even reach the browser. It's a reverse engineered program of the popular Proximitron in the early 2000s.

There's also a recent separate effort called Proximitron Reborn, but I think the author of that hasn't release the source she promised yet. On going discussion of it is on the ancient prxbx forum here: https://prxbx.com/forums/showthread.php?tid=2331

**Comment by mikhoul** [comment]
Just to let know user that even the security guys at Mozilla seem to think it's a good Idea to weaken the API for extensions: https://old.reddit.com/r/firefox/comments/aithmh/raymond_hill_creator_of_ublock_origin_ubo_and/eerce78/

Firefox seem to be just an inferior "Chrome clone" right now with no real differentiation and no advantage sadly.

Keeping a strong and better API than Chromium could have been their Hail Mary but hey it's Mozilla as usual AKA stupid.

Regards :octocat:

**Comment by jspenguin2017** [comment]
@onmyouji 
Yes, I'm aware that a browser based on Electron is hard to secure, but so is extending a browser using native code. 

As I stated before, I have no plan in pursuing this option. Thanks for the information though.

**Comment by jspenguin2017** [comment]
> @nl255
> Are you saying that websites will simply block Firefox users entirely if Firefox insists on keeping the existing webRequest API?

That's a possibility. 

> For this to work it would need to support HTTPS interception/rewriting which while it can be done opens up a huge can of worms security wise, especially if it is not done properly.

I'm aware of it, that's why I want to use a well established project to handle the HTTPS repacking part. 

> It would also require having a companion GUI app

We'll worry about that part later. That's going to be easy compared to the rest. 

> Even worse, on Android by default apps will not trust user installed root certificates

I thought that's only for apps that use certificate pinning. Either way, let's get it working on Linux and Windows first, we'll worry about Android later.

**Comment by mikhoul** [comment]
Something that passed under the radar this autumn Mozilla want to follow V3 manifest too: https://blog.mozilla.org/addons/2018/10/26/firefox-chrome-and-the-future-of-trustworthy-extensions/

So the comment of Mozilla on Reddit is logical: https://old.reddit.com/r/firefox/comments/aithmh/raymond_hill_creator_of_ublock_origin_ubo_and/eerce78

**Comment by jspenguin2017** [comment]
> @mikhoul
> it'S possible and Brave seem to already use this strategy

I knew it's possible, and I knew Brave is doing it. That does not change the fact that coding in native code comes with risks, especially when the native code interfaces with untrusted inputs (web traffic).

**Comment by mikhoul** [comment]
[**@jspenguin2017**](https://github.com/jspenguin2017) commented on [23 janv. 2019 à 22:22 UTC−5](https://github.com/NanoAdblocker/NanoCore/issues/238#issuecomment-457052621 "2019-01-24T03:22:23Z - Replied by Github Reply Comments"):
> > [@mikhoul](https://github.com/mikhoul)  
> > 
> > it'S possible and Brave seem to already use this strategy
> 
> I knew it's possible, and I knew Brave is doing it. That does not change the fact that coding in native code comes with risks, especially when the native code interfaces with untrusted inputs (web traffic).
> 
> > [@joey04](https://github.com/joey04) (from other thread)  
> > 
> > But a browser fork is a _massive_ effort.
> 
> It doesn't have to be. As you mentioned, the hard part is already taken care of by major tech companies and their paid developers. All we need is to insert our code at the right place and keep the fork up to date with upstream.
> 
> There are 2 problems though:
> 
> *   The code that we insert probably have to be native code, which can introduce new vulnerabilities
> *   It's hard to get WideVine (DRM) working, also, Chromium depends on Google API a lot, you can get the API keys for your personal use fairly easily, but if you distribute your build the quota will run out pretty quickly
> 
> I won't say it's impossible, but it's hard enough that we want to explore other possibilities first. It's not necessarily a massive effort, but it'll probably be a massive paperwork.

I'm not sure on how hard it is for Widevine since https://chromium.woolyss.com/ have all the version build in almost all flavors quickly since many years with Widevine, if it was complicated they would not distribute it so widely.

But I understand that other API have quotas limit or a restricted  but I never suffered from it but like I said I don't use the sync.

I  browsed quickly https://chromium.woolyss.com/#google-api-keys and found it's relatively easy to install API key if you really need it, it seem it's a thing you do once and forget about it. 😏

I would really like a Chromium fork for power-users on desktop with a more compact GUI with less padding and above all with Nano integrated. 

 I know it's not the subject but take a look at [CentBrowser](https://www.centbrowser.com) they added lot of features for power-users. but he is closed source that's the downside but I use it with other forks.

Tomorrow I will ask to the CentBrowser team what is their take about V3 Manifest and restriction, they are a small team (~3 people) but they are responsive and open and I will report.

Regards :octocat:

**Comment by nl255** [comment]
@jspenguin2017 



>    Even worse, on Android by default apps will not trust user installed root certificates
>
>I thought that's only for apps that use certificate pinning. Either way, let's get it working for desktop Linux and Windows first, we'll worry about Android later.

Nope, it applies to  [Nougat and above](https://blog.jeroenhd.nl/article/android-7-nougat-and-certificate-authorities).  Even worse, there is no reason to think that Chromium and Firefox won't do something similar as they have already started heading that way with DNS over HTTPS (which also conveniently gives google access to all your DNS requests even if you run your own DNS server).  So you might end up having to make either a fork or possibly a program loader (since DLL injection is blocked that leaves a loader program that patches the browser in memory) but of course with a loader they could fight back by using various tricks to make pattern based smartpatching fail possibly even including a Denuvo style custom virtual machine.

.

**Comment by elypter** [comment]
even with all the harmful decisions firefox is still the better choice. i think it is more likely that firefox at least keeps the api somewhat better than chrome. and even if not, there are also firefox forks like for example https://github.com/intika/Librefox/issues which aims to create an always up to date version with an automated build system

**Comment by nl255** [comment]
Even with the potential issues mentioned earlier I would think the native, standalone app would be the best choice as it will be far easier to maintain a fork that merely makes SSL interception easier than it would be to reimplement the existing webRequest API especially if the browser code ends up being refactored to make it harder to do so as short of going proprietary or heavy code obfuscation/encryption it will always be easier to comment out a few functions regarding SSL certificate checks or make them always return "true" even if the check failed than it will be to keep the webRequest API working.

**Comment by mikhoul** [comment]
> standalone app would be the best choice as it will be far easier to maintain a fork

I'm not sure about it at the end you will need an extension to be able to select element on the fly and create filters, even [AdGuard](https://adguard.com/en/faq.html) that use a stand alone app need to have an extension and the dev already stated the such change in the API will hurt AdGuard.

**Comment by joey04** [comment]
>>    @joey04 (from other thread)
>>    But a browser fork is a massive effort.

> It doesn't have to be. As you mentioned, the hard part is already taken care of by major tech companies and their paid developers. All we need is to insert our code at the right place and keep the fork up to date with upstream.

"All we need" eh?  :-)

Kidding aside, I understand what you mean. Perhaps _time-consuming_ is a more accurate term from a Chromium fork with the level of custom code we'd want for robust content filtering.

Just building a Chromium pull alone can take an hour or more depending on your CPU. And in the long run, there's the possibility of Google changing something internally that would impact an integrated custom filter engine. That could be a nightmare to deal with.

So I'm still of the opinion that modifying the [Brave engine](https://github.com/brave/ad-block), which they generously share as MPL, is the best approach. It's already integrated into Chromium, and if Google breaks that interface in some way, the Brave devs will have to deal with it.

As for the other proposed idea, intercepting security certs sounds too much like SuperFish. I don't think I'd be comfortable with a benevolent FOSS one.

**Comment by nl255** [comment]
>I'm not sure about it at the end you will need an extension to be able to select element on the fly and create filters, even AdGuard that use a stand alone app need to have an extension and the dev already stated the such change in the API will hurt AdGuard.

It could be done without an extension, it would just be uglier and depending on how it is done would likely require customization for each browser and major browser revision. Most browsers have an inspect element right click function which will bring up the source for the element in the developer tools which could then be OCRed via screen grab and compared to the page source captured by the standalone app.  Or possibly by injecting custom javascript (AJAX?) code that communicates with a locally running server to get around the lack of sufficient APIs.  Or worst case the user would have to find the element using the inspect element and create the filter on their own (meaning only advanced users who know html would be able to block elements on their own).  Though that could probably be worked around by using a fork that has dll injection detection/prevention disabled combined with an injected dll (or equivalent for other operating systems) meaning you would have to install Nanofox or Nanonium to be able to create your own element filters through a point and click interface.

**Comment by mikhoul** [comment]
> It could be done without an extension, it would just be uglier and depending on how it is done would likely require customization for each browser and major browser revision. Most browsers have an inspect element right click function which will bring up the source for the element in the developer tools which could then be OCRed via screen grab and compared to the page source captured by the standalone app.

At his point it would be **lot more effort and time** that forking Chromium.... 😵

Also most users are not interested in opening Dev tool to hide an element, it's not convenient at all. I do that myself but only for really complex  ads, 99% of the time I use  the selector mode of Nano.

The goal of @jspenguin2017  seem to be to lower the load as much as humanly possible so i'm really not sure. 

Also @joey04 have a really good argument i**t's really not everybody that would be willing to install a root certificate** for an Adblocker I don't even trust AV to install a certificate.

Regards :octocat:

**Comment by jspenguin2017** [comment]
Sorry guys, I'm quite short on time these days. 
I'll respond to your comments eventually, but for now, don't get too worked up on this. 

Keep voicing your concerns and displeasure to Google via media outlets to apply pressure. Don't spam technical threads though, that can backfire.

**Comment by jspenguin2017** [comment]
> mikhoul
> if it was complicated they would not distribute it so widely

I would guess the paperwork is a one-time thing. So even if it's really complicated, once you get the it through you can distribute it widely without issues. 

> it's relatively easy to install API key

It's not bad, but I'll still consider it to be a problem if everyone have to do it.

**Comment by jspenguin2017** [comment]
> @nl255
> DNS over HTTPS

Since we are going to repack HTTPS, there's no reason that we can't inspect and alter DoH responses. 

Desktop browsers do accept custom certificates and I don't think that's going to change anytime soon. There is no need to patch browsers in any way. 

If you're talking about Android, then that's another story, honestly I'm not too sure what to do.

**Comment by jspenguin2017** [comment]
> @mikhoul
> even AdGuard that use a stand alone app need to have an extension

Yea, it's probably going to be kind of tricky to implement cosmetic filters with just a native app. Script snippets injection is probably better handled by extensions too. 

Parsing and altering server response is going to be a heavy task, not to mention CSP can sometimes be in the way.

**Comment by jspenguin2017** [comment]
> @joey04
> Perhaps *time-consuming* is a more accurate term for a Chromium fork

By "all we need" I meant that it's kind of a one-time task. Once everything is in the right place, the maintenance task should be minimal. 

> there's the possibility of Google changing something internally that would impact an integrated custom filter engine

If done right, I don't think it'll be a problem. We can hook into `declarativeNetRequest` and process filters on the spot synchronously with native code, as long as there's some sort of filtering system left, we should be good. 

> I don't think I'd be comfortable with a benevolent FOSS one.

But you're fine with a custom browser? 
I think a custom browser is more risk than a MITM proxy. I plan on writing the proxy in JavaScript or Python, which are memory safe. So as long as the code is not malicious, you'll likely be safe. 
To patch a browser, we'll need native code, which can open vulnerabilities that can be exploited by websites you visit.

**Comment by jspenguin2017** [comment]
> @nl255
> It could be done without an extension

Considering the native app can control everything the browser receives, it can definitely be done without an extension. However, it'll be horrifyingly hard, as we'll have to patch websites on a case by case basis.

**Comment by jspenguin2017** [comment]
> @mikhoul
> I don't even trust AV to install a certificate

Considering AVs usually have companion kernel drivers, you're probably better just not using an AV.

**Comment by mikhoul** [comment]
@jspenguin2017  Thanks for and I understand you can be short in time so no worry.

Right now the Brave team seem really be interested to keep the webrequest API  as it is with without neutering it.  They stated it clearly on Reddit, it seem to be the same for CentBrowser so I'm pretty confident that at least a fork will keep the webrequest API  retro-compatible with Manifest V2 as it is today.

The main problem right now is that all those dev and everybody don't really know  what will be in Manifest V3 at the end so it's hard  for them and everybody to begin to plans on how to keep/protect the API as it is and see how to do it technically speaking.

Also personally I could be wrong but I don't think in the long run Mozilla/Firefox will keep the webrequest API as it is today, they will follow Google path "in the name of security" ... they already done it with xul and they promised to port old APIs and after the said no it would breach security. In short I don't trust Mozilla at all sadly.

I will follow closely other Chromium forks and report here informations about what they will do with the webrequest API once Google will have set V3 in stone.

Wishing you a nice day ! 🌞

**Comment by joey04** [comment]
> Test out your ideas and find potential issues, that's going to benefit everyone here.

We'll have to agree to disagree on whether a browser or MITM approach is best, but I certainly agree that at this preliminary stage it's good that different people are exploring various options. I've decided to keep an open mind because there are too many TBDs at this point to put all my eggs in one basket.

I've starting looking into Chromium options a bit more and made a lengthy post on a [ungoogled-chromium thread](https://github.com/Eloston/ungoogled-chromium/issues/662). I won't be posting more about it here.

**Comment by jspenguin2017** [comment]
> @mikhoul
> I'm pretty confident that at least a fork will keep the webrequest API

That's great news!

> @joey04
> We'll have to agree to disagree on whether a browser or MITM approach is best

As you said, don't put all eggs in one basket. It's actually good if we disagree, more diversity means more resilience in time of crisis, if we end up starting 2 different projects, we'll have more things to work with next time something unexpected comes up. 

>  ungoogled-chromium

AFAIK, that project aim at implementing Chromium features that depends on Google API. Chromium can run without Google API just fine, it will just be missing some features. Overall, I think that project is rather unrelated to what we are trying to do.

**Comment by jspenguin2017** [comment]
> @joey04
> Just building a Chromium pull alone can take an hour or more depending on your CPU

You need at least 16GB of RAM, for CPU, an i5 is enough. You also need 100GB of hard disk space to store build cache. So yea, you need a decent computer to build Chromium, worst case you can build overnight. 

Note that no matter how bad your computer is, as long as it is able to produce the first build, subsequent builds should be fairly fast due to build cache.

**Comment by xlollomanx** [comment]
@mikhoul seems like firefox confirmed that they will not follow google mind about adblockers https://www.reddit.com/r/firefox/comments/ajlhip/firefox_product_manager_on_twitter_firefox_isnt/

**Comment by mikhoul** [comment]
> @mikhoul just to add more infos, seems like firefox confirmed that they will not follow google mind about adblockers

Sadly it's just a Mozilla employee that give his personal opinion and nothing more, the other guy that made the first comment about the API is in charge of the security at Mozilla/Firefox so his comment are weighting lot more since he is part of the team that decide what is secure or not.

Don't get me wrong I would be really happy if Mozilla would not neuter any API with V3, it would make one more alternative for everybody to use.

_______________________________________________________________

On another point like @gorhill  has said there is another serious issue with the V3 Manifest with the replacement of background pages: https://github.com/uBlockOrigin/uBlock-issues/issues/338#issuecomment-457609180 . Its also a serous issue for more than 50% of existing extension IMO.

Regards :octocat:

**Comment by FatOrangutan** [comment]
>Sadly it's just a Mozilla employee that give his personal opinion

He is a Firefox product manager and "responsible for the browser roadmap."

>the other guy that made the first comment about the API is in charge of the security at Mozilla/Firefox so his comment are weighting lot more since he is part of the team that decide what is secure or not

Um, no. Even otherwise, the PM said "not happening," while the sec guy said "not bad." The latter isn't a statement of intent, while the former is.

**Comment by joey04** [comment]
Interesting wrinkle. The tweet by Asa Dotzler, product manager at Mozilla, reads: "What makes you say Firefox is doing anything. This announcement was about Chrome, not Firefox."

I wouldn't read too much into it, and actually it's not that encouraging. I agree with the top-rated comment on the Reddit thread:

> that's not the outright denial I would like to hear from Mozilla. Their marketing group should be working overtime getting the message out that Firefox will not be crippling ad-blockers. Mozilla is being handed a great opportunity, they should use it to their advantage.

There's still hope Mozilla doesn't screw the pooch here, but I'm skeptical they will in the long run. I don't trust them.

For now, ESR remains my primary browser. I'm of course exploring other options too.

**Comment by elypter** [comment]
@joey04 

Librfox looks interesting. I hope it can get beyond the frst steps. It aims to create a security and freedom of choice focused fork that is always up to date with the current version by automating the patching & building process

**Comment by jspenguin2017** [comment]
I'm starting to prototype a native app that emulates WebExtension APIs, let's see how it goes.

**Comment by mikhoul** [comment]
> I'm starting to prototype a native app that emulates WebExtension APIs, let's see how it goes.

It's Good thing and let us know how the progress goes ! 

Regards :octocat:

**Comment by nl255** [comment]
@jspenguin2017

By native app do you mean a standalone proxy program or do you mean native code intended to be added to the browser itself similar to what Brave does with it's ad blocking.  Also, you might want to take a look at the mitmproxy project if you haven't already.

**Comment by jspenguin2017** [comment]
@nl255 I mean a proxy server.

**Comment by joey04** [comment]
> @elypter 
> the key question is how well the patching and building can be automated
> it all depends on intelligent merging and development tools.

What does Librefox use that other projects don't?

I've not used Mercurial (or even Git in a non-trivial way) but I was under the impression it could be readily integrated with your Diff tool of choice.

**Comment by joey04** [comment]
Kinda off-topic, but does anyone know the answer to [my question on the other thread](https://github.com/Eloston/ungoogled-chromium/issues/662#issuecomment-457879639)?

I've been purposely keeping my long posts separate because this thread should now focus on @jspenguin2017's new project. But any help on this matter would be appreciated.

**Comment by jspenguin2017** [comment]
> @joey04
> It just occurred to me, perhaps there's a way to do Chromium development without need of my own build machine.

You pretty much either have to build on your own device or run a VM on AWS. You'll have some fun time developing if you rely on CI servers for building. 

I don't think Travis will work because the spec of their VMs are pretty low. 
See: https://docs.travis-ci.com/user/reference/overview/

**Comment by jspenguin2017** [comment]
Any recommendation for Public Suffix library for TypeScript?

**Comment by joey04** [comment]
@jspenguin2017 
Thanks for the CI info. I don't have experience setting up remote VMs like that, and since GitHub requires running on Google's cloud for it, it would cost me some money even to try. I'm not willing for pay-to-hack yet.
Perhaps someone else will have a feasible suggestion, though.

**Comment by jspenguin2017** [comment]
> @joey04
> I don't have experience setting up remote VMs like that

Try this one, it worked for me. 
https://www.digitalocean.com/community/tutorials/how-to-install-and-configure-vnc-on-debian-9

**Comment by jspenguin2017** [comment]
So far so good. I'm not sure how hard it is to implement HTML filtering though. Also WebSocket will be a pain. 

![image](https://user-images.githubusercontent.com/7283682/51796784-a73b7680-21b5-11e9-9cbf-0d4587ce0bbf.png)

**Comment by nl255** [comment]
@jspenguin2017 
It is my understanding that Chrome/Chromium simply use the system wide proxy settings by default and can be forced to use a specific proxy server on all platforms using the command line both of which is highly unlikely to go away even if they remove the ability to access those settings via extension.

**Comment by jspenguin2017** [comment]
The core part of the native app is mostly ready, only thing missing is WebSocket. 

> @nl255

OK, thanks for the information.

**Comment by jspenguin2017** [comment]
> @nl255
> I don't know what language you are using

I'm using TypeScript. 

I decided to not use `mitmproxy` because it's in Python, and uBO requires a JavaScript runtime.

**Comment by jspenguin2017** [comment]
Just a quick update, a basic inspection engine is mostly ready. It can filter by URL and modify headers sent from browser for now, I'll expand it later. I'm planning on adding ability to modify headers sent from server and content sent over non-WebSocket connections. 

I still need to do some testing to make sure everything is working smoothly. The only technical hurdle left is inspecting and modifying WebSocket frames, but that's not really a priority for now.

**Comment by mikhoul** [comment]
Good News:  **Google backtracks on Chrome modifications that would have crippled ad blockers**
https://www.zdnet.com/article/google-backtracks-on-chrome-modifications-that-would-have-crippled-ad-blockers/

@jspenguin2017  Are you going to continue to develop the standalone extension in parallel ? 🤔

Regards ! :octocat: 

*Note* :  I will continue to follow the Manifest v3 to see if they will also remove the background page without any real alternative.

**Comment by mikhoul** [comment]
@jspenguin2017 ,  After reading more and looked at the fact  directly from Google it seem that the ZDNet article is really misleading. The author didn't do his research of it's some PR from Google to minimize the bad look on them.... :disappointed: 

Nothing seem to have changed they Google to remove the webrequest API, it's sad to see such poor journalism.


I was thinking it was some last minute statement from Google but the joy was shortlived...

**Comment by jspenguin2017** [comment]
> Are you going to continue to develop the standalone extension in parallel ?

For now yes. The new APIs are going to be pretty easy to use, The [current prototype from Adblock Plus](https://gitlab.com/eyeo/adblockplus/adblockpluschrome/blob/convert-declarative-net-request-rules/build/convert-declarative-net-request-rules.js) is only a few hundred lines. 

> it's some PR from Google to minimize the bad look on them....

Not really. Assuming they actually implement what they promised, we should have most APIs that we need. The issues that I can think of right now are: 
- The `important` option
- The `csp` option
- The lack of DOM in the extension process
- The new hosts permission restrictions

The `important` and `csp` options have workarounds (as explained above). The background DOM can be emulated with [a library](https://github.com/jsdom/jsdom). As for the new hosts permission restrictions, if I understood correctly, it just means longer review and maybe one more setup step.

**Comment by nl255** [comment]
@jspenguin2017 

Well keep in mind that even if they deliver everything they promised (which is unlikely) it will still be impossible to block "acceptable ads" via browser extension since it will be up to the browser to decide whether or not a block request will be honored.  Not to mention that thanks to Signed HTTP Exchanges domain based blocking (whether via browser extension or even hosts file/DNS based blocking) will become useless as well.

**Comment by jspenguin2017** [comment]
> it will be up to the browser to decide whether or not a block request will be honored

We'll find out pretty quickly if Google sneaked in some whitelists. It's not like extensions today can work correctly without the browser being honest.

**Comment by nl255** [comment]
Is there any new news on this?  Since it appears that google is standing firm on their Manifest V3 changes though there will be an exception for paid enterprise users.

**Comment by mikhoul** [comment]
At this point we don't know if Chromium fork will be able to enable the policy that permit full adblockers.

My guess is that those restrictive policy in the V3 Manifest will be only implanted in chrome and not in chromium.

Worst come to worst I will migrate back to Firefox. :octocat:

**Comment by jspenguin2017** [comment]
> Is there any new news on this?

Not much, I have my prototype built for option 3 (see posts at the beginning) but I have yet to decide whether I want to carry through with the idea. 

> we don't know if Chromium forks will be able to enable the policy

Considering the code for the current `webRequest` is already public, there isn't a reasonable way to stop forks from using it.

**Comment by mikhoul** [comment]
Brave already intend officially to make webRequest backward compatible: https://twitter.com/BrendanEich/status/1134141335881912320?

I've asked also to Brendan if they also intend to let extension use remote code to allow Userscripts/Userstyles mangers to exist....  I'm pretty sure they will allow them too.

**Comment by KonoromiHimaries** [comment]
> Wexond is an extensible and `privacy-focused` web browser with a totally different user experience, built on top of `Electron``

https://github.com/wexond/wexond

**Comment by kiwibrowser** [comment]
Feedback from Kiwi Browser (for Android),

Nano Adblocker is the content filter extension that we recommend the most to new users (even more than uBlock Origin) so we'll keep compatibility with any v2 API you'd like to use @jspenguin2017

**Comment by mikhoul** [comment]
> Nano Adblocker is the content filter extension that we recommend the most to new users (even more than uBlock Origin) so we'll keep compatibility with any v2 API you'd like to use @jspenguin2017

Great News  @kiwibrowser  !!! You almost read on my mind  since I was about asking you if you would keep compatibility for the webRequest API since I use Kiwi and install it on all TV-Box/phone sI use and for my friends too.

Do you intend to allow remote code also for extensions like Tampermonkey and for the others userscripts/userstyles managers  ?  🤔

Wishing you a great week-end ! :sun_with_face:

**Comment by kiwibrowser** [comment]
Yes

**Comment by jawz101** [comment]
I was wondering if there is a blocking mechanism that could distrust SSL certs.  As they prevent extensions from tampering w/ traffic, use Chrome data savers to compress by sending your traffic through them, use the async dns resolver pref to handle dns resolutions within the browser- what Chrome doesn't manage is the local SSL certificate store.  And since most stuff is SSL you'd just distrust the certs for servers you don't like in your local cert store.

**Comment by jspenguin2017** [comment]
> I was wondering if there is a blocking mechanism that could distrust SSL certs.

That's way too overcomplicated for what it achieves. The granularity of certificate-based filtering is at the domain level at best, so DNS-based filtering solutions will probably have better granularity while being way more mature.

**Comment by mikhoul** [comment]
An update here: https://blog.chromium.org/2019/06/web-request-and-declarative-net-request.html

I did not read it carefully but it seem enterprise will be able to by pass with policy so most most Chromium fork targeted to power users should be able to bypass the webRequest restrictions easily.

But on my side I will probably migrate to Brave for my desktop browser since they will fix the other issues/restrictions that Manifest V3 introduce, [it's not only the webRequest API that will be affected sadly](https://9to5google.com/2019/01/29/chrome-manifest-v3-tampermonkey/).

For Android I will keep Kiwi Browser since it will allow me to use the same extensions/API that I use today.

**Comment by mserajnik** [comment]
I've made the switch to Brave on all my devices (macOS/iOS/Win 10) recently and while there's some things that are still inferior/missing in comparison to Chrome, for me personally, it made more sense than starting to use Firefox (where I was missing extensions and didn't like the dev tools as much).

Plus, I support Brave's content monetization ideas; they are not a perfect solution, but any alternatives to privacy-invasive ads/trackers are welcome.

So, I hope that you will continue developing Nano for Chromium-based browsers like Brave that will bypass the V3 changes, regardless of what happens with Chrome.

**Comment by mikhoul** [comment]
[**@mserajnik**](https://github.com/mserajnik) commented on [20 juin 2019 à 14:42 UTC−4](https://github.com/NanoAdblocker/NanoCore/issues/238#issuecomment-504137910 "2019-06-20T18:42:02Z - Replied by Github Reply Comments"):

> So, I hope that you will continue developing Nano for Chromium-based browsers like Brave that will bypass the V3 changes, regardless of what happens with Chrome.

From @jspenguin2017  mouth: 

> Considering the code for the current webRequest is already public, there isn't a reasonable way to stop forks from using it.

Brave is not the only one who stated that they will keep alive webRequest API Vivaldi too: https://www.theinquirer.net/inquirer/news/3077032/vivaldi-promises-to-restore-third-party-ad-blocking-to-chromium

But we need more than webRequest API we need [other API/methods that will be deprecated too](https://9to5google.com/2019/01/29/chrome-manifest-v3-tampermonkey/), but Brave said that he will keep them too along with the webRequest API.

At the end of the summer I will switch to Brave or Opera too for my desktop use, worst come to worst it will be Firefox.

**Comment by jspenguin2017** [comment]
> we need other API/methods that will be deprecated too

Tampermonkey needs to execute remote code, which Google will no longer allow. Note that Mozilla never allowed executing remote code, but they made exceptions for those popular userscript extensions. 

I'm not aware of other APIs that we need being removed at this point.

**Comment by mikhoul** [comment]
[**@jspenguin2017**](https://github.com/jspenguin2017) commented on [20 juin 2019 à 20:58 UTC−4](https://github.com/NanoAdblocker/NanoCore/issues/238#issuecomment-504250051 "2019-06-21T00:58:33Z - Replied by Github Reply Comments"):
> > we need other API/methods that will be deprecated too
> 
> Tampermonkey needs to execute remote code, which Google will no longer allow. 
> I'm not aware of any other APIs that we need being removed at this point.

The founder of Brave told that he will keep the permission to use remote code so extensions like Userstyles Managers & Userscripts managers will be able to exist, I'm pretty sure it will be the same for Vivaldi but I'm not 100% sure for Vivaldi.

The other things that many devs complaint is the removal of "background page" but I'm not sure if it will be replaced or not with others alternatives or not

I remember that Manifest V1 was completely removed from Chrome code only 12-18 months ago so we should have at least  12-24 months after the approval of V3 before they remove it completely from Chrome.

**Comment by jspenguin2017** [comment]
> many devs complaint is the removal of "background page"

Nano does not strictly need a background "page", it is a good-to-have, but not required. The new background service worker should be enough. 

The only time Nano uses the background page renderer is to validate CSS selectors, but there are many alternatives: 
1. Borrow the renderer from one of the open pages
2. Create a temporary tab / window when needed
3. Ship a library / polyfill
4. Use a web service

**Comment by nl255** [comment]
@jspenguin2017  I still think that the standalone app is probably the way to go especially if uMatrix (NanoMatrix?) and Tampermonkey/Violentmonkey like functionality can be added later as well.

After all, even without any special effort by google to complicate things it will be increasingly difficult for the various forks to maintain said compatibility as well as having to keep the extension updated on half a dozen to a dozen different addon/extension "web stores" assuming said fork even has their own.

**Comment by mikhoul** [comment]
[**@nl255**](https://github.com/nl255) commented on [21 juin 2019 à 11:43 UTC−4](https://github.com/NanoAdblocker/NanoCore/issues/238#issuecomment-504473033 "2019-06-21T15:43:15Z - Replied by Github Reply Comments"):
> [@jspenguin2017](https://github.com/jspenguin2017) I still think that the standalone app is probably the way to go especially if uMatrix (NanoMatrix?) and Tampermonkey/Violentmonkey like functionality can be added later as well.
> 
> After all, even without any special effort by google to complicate things it will be increasingly difficult for the various forks to maintain said compatibility as well as having to keep the extension updated on half a dozen to a dozen different addon/extension "web stores" assuming said fork even has their own.

I don,t agree here, it would require a colossal sum of time and it would be really hard to maintain, All in one extension are the worst to maintain that's why UBO & Umatrix are 2 different extensions adding Tampermonkey/Stylus to the mix would be foolish IMO.

Moreover with a standalone approach you will lose lot of granularity and ability.

There is no real problem since 2 major forks (at least) are committed to keep the old API and extensions are far more "easy" to keep updated compared to a Frankenstein All-in-one standalone app :wink:

**Comment by nl255** [comment]
@mikhoul  Not compared to maintaining multiple standalone proxies for each extension and trying to get them to play nice when daisy chaining them together.    As for those forks that is nice in the short term but it will become increasingly difficult to maintain those old APIs as time goes on.  Not to mention that with Brave and other forks besides Opera how exactly are you going to get the extension since you won't be able to get it from google's "web store" anymore (unless they disable extension signing completely like Kiwi does)?

**Comment by jspenguin2017** [comment]
> and it would be really hard to maintain

Depends on how the proxy engine is designed, it can be designed to be extensible. 

> Moreover with a standalone approach you will lose lot of granularity and ability.

I don't believe a MITM proxy has less granularity. 
In terms of ability, blocking pop ups might be a bit hard. Nothing showstopping though.

**Comment by jspenguin2017** [comment]
> it will be increasingly difficult for the various forks to maintain said compatibility

If you're talking about legacy manifest v2 extensions, then yes. But for manifest v3 extensions that use blocking `webRequest` API, then probably not, since Chrome will still allow enterprise installations to use it. 

> having to keep the extension updated on half a dozen to a dozen different addon/extension "web stores"

Yes, that is certainly a problem. Especially when some stores don't support publishing from command line.

**Comment by jspenguin2017** [comment]
I know it's been a while, but I have yet to make up my mind. I have been observing the situation develop though, and here's my update on the "3 options" I listed at the very beginning. 

* The new API is still underpowered, but becomes more and more viable with each update. I'll investigate deeper after another few updates to evaluate its viability. 

* As for forking Chromium, this is no longer needed. Instead, this option is replaced by moving to a browser that supports blocking `webRequest` in normal installation. This comes with its own issues though, like how to quickly publish updates to many extension stores. 

* I have explored making an native app, and I made an initial prototype which confirmed the theory. However, making a fully functional native app is a non-trivial challenge, and I'm not sure if I'll have enough time to finish it in time. 

Here are the important things to watch next. 

* Microsoft's move. As the new Edge is getting closer and closer to stable release, Microsoft's response to manifest v3 is probably going to be the most impactful decision on this whole situation. 

* Upstream's move. Depending on what upstream does once Google makes the final decision, we may need to adjust our plan to minimize headache during the transition.

**Comment by mikhoul** [comment]
[**@nl255**](https://github.com/nl255) commented on [21 juin 2019 à 13:29 UTC−4](https://github.com/NanoAdblocker/NanoCore/issues/238#issuecomment-504506728 "2019-06-21T17:29:14Z - Replied by Github Reply Comments"):
> [@mikhoul](https://github.com/mikhoul) Not compared to maintaining multiple standalone proxies for each extension and trying to get them to play nice when daisy chaining them together. As for those forks that is nice in the short term but it will become increasingly difficult to maintain those old APIs as time goes on. Not to mention that with Brave and other forks besides Opera how exactly are you going to get the extension since you won't be able to get it from google's "web store" anymore (unless they disable extension signing completely like Kiwi does)?


Not really since Brave already maintain its own code for their own addblocking and for the webrequest API like  @jspenguin2017  already stated the code will be in Chrome/Chromium and even it it was removed it's easy for Brave/Opera to maintain different API they both have already of custom API/mechanism.

> how exactly are you going to get the extension since you won't be able to get it from google's "web store" anymore (unless they disable extension signing completely like Kiwi does)?

There is many way for me I just side-load in dev mode for the casual user Opera have already it's own store and it will probably be the same for Brave.  With Microsoft  browser it's the same they have thir own store for extension and you can choose from the Chrome store of the Microsoft store.

Brave is already ready to have their own store for the extension that would like to use their specific API/features: https://github.com/brave-experiments/brave-extension-store

Also there is  nothing complicated to keep an extension updated for the user the update is automatic  and for the Dev I'm pretty sure Brave and Opera will make the publishing process easy for the uploading/updating  for the extensions that will  take advantage of their particularity.

Don't forget that Firefox could also be an alternative again for power users IF they don't follow the path of Google for V3 API...  However I will not hold my breath for Mozilla doing the "good" thing... I have been deceived to many times by their decisions.


**BTW**: Chrome team today stated again they want global limit for rules not per extensions which will limit a lot many extensions.

**Comment by nl255** [comment]
@jspenguin2017 As I understand it, the blocking webRequest API is only available in the paid Enterprise version and thus the code for it won't be present in the Chromium source code (i.e. the patch to enable/implement it will be closed source).  Otherwise it would probably be fairly easy to patch Chrome or write a Chrome loader to enable it for all users*.

As for the native app, if you do decide to go that way you might want to contact the EFF as their privacy badger addon has the same issues so perhaps they can help finish it so they can use the same code as well.

*While Chrome does have some dll injection protection it probably doesn't have anything even on the level of Shiva to prevent in memory or even on disk patching much less Denuvo level.  You could probably even use smart patching (i.e. pattern based rather than offset based patching) to make it work with multiple versions.

**Comment by jspenguin2017** [comment]
> it's easy for Brave/Opera to maintain different API

If they roll their own APIs, I would count that as a fail. That would quickly become too fragmented. 

> I'm pretty sure Brave and Opera will make the publishing process easy

I only heard about complains about the Opera store. 

> Chrome team today stated again they want global limit

As I understood, the per-extension limit would be lifted once the global limit is implemented. Note that the rules limit in Safari is 50k, and some people / newspapers consider Safari to be an adblocker-friendly browser.

**Comment by jspenguin2017** [comment]
> the blocking webRequest API is only available in the paid Enterprise version

While it is possible for Google to strip the code from Chromium and only ship it in Chrome, I think you can get the enterprise version for free. 

Even if they do strip the code, I don't think they'll do anything to make it painful to re-add it as they would be stepping on their own toes.

**Comment by mikhoul** [comment]
One more desktop browser will officially support the old webRequest API: https://www.zdnet.com/article/opera-brave-vivaldi-to-ignore-chromes-anti-ad-blocker-changes-despite-shared-codebase/  :smile:

**Comment by jspenguin2017** [comment]
RegExp filters are being implemented: https://docs.google.com/document/d/1foL79VwByaVeUwYPBfZeqyC2SYcAawhjjH05eqHqW_E/edit

**Comment by mikhoul** [comment]
Hi @jspenguin2017 ,

Not precisely related to Manifest V3 but it's in the same vein with Google attempt to restrict ad-blockers  ...

From this Issue in UBO https://old.reddit.com/r/uBlockOrigin/comments/dgncga/dev_build_1225rc1_rejected_from_chrome_web_store/ I understand that UBO will possibly be no longer available from the CWS and I fear the same thing will  happen to Nano.

What is your take on this @jspenguin2017 ?  🤔

Regards !

**Comment by jspenguin2017** [comment]
@mikhoul Sometimes false positive happen, it usually gets resolved after a resubmission or an email. I don't think it's a problem.

**Comment by okiehsch** [comment]
>  usually gets resolved after a resubmission or an email.

https://github.com/uBlockOrigin/uBlock-issues/issues/745#issuecomment-539962215

**Comment by mikhoul** [comment]
> uBlockOrigin/uBlock-issues#745 (comment)

Gorhill have a rigid personality and did not change anything in the code when he resubmitted so it's expected to receive the same results. Just changing some code "cosmetically" could have resulted in the the extension to have passed instead of being rejected.

Gorhill has many quality (very good developer, focused, hard worker)  but he is not flexible and he is a little bit in his bubble, he did not even try to contact Google by other mean than answering to the email, he just marked the issue as "wontfix".

Anyway right now Google is looking in the issue:  https://old.reddit.com/r/chrome/comments/dgoymg/warning_ubo_ublock_origin_will_possibly_be/f3e4fi6/

![image](https://user-images.githubusercontent.com/5678673/66704030-90f32d80-ece6-11e9-8a95-017ad8364db4.png)

**Comment by okiehsch** [comment]
[1.22.5rc2](https://chrome.google.com/webstore/detail/ublock-origin-development/cgbcahbpdhpcegmbfconppldiemgcoii) has been approved in the CWS.

> did not change anything in the code when he resubmitted
After they rejected `1.22.5rc2` he submitted `1.22.5rc2` which was also rejected.
Here are the code changes https://github.com/gorhill/uBlock/compare/1.22.5rc1...1.22.5rc2.

**Comment by mikhoul** [comment]
> 1.22.5rc2 has been approved in the CWS.

Good news so it was just some kind of mistake of a false positive like @jspenguin2017  was saying.

**Comment by jspenguin2017** [comment]
It's likely caught in some automatic filter, usually sending an email to request a human review corrects the problem.

**Comment by okiehsch** [comment]
> Good news so it was just some kind of mistake of a false positive

I never thought it was some nefarious scheme to get rid of `uBO` but I find it worrying that after nearly ten years of CWS they still have a system in place that makes it very difficult for developers to resolve issues with the automatic filter.
See the links at https://github.com/uBlockOrigin/uBlock-issues/issues/745#issuecomment-541343439 for more examples.

**Comment by jspenguin2017** [comment]
I feel that WebStore is easier to use (for developers) than Microsoft Store and AMO. Although it looks ugly, it is blazing fast. Microsoft Store is definitely the worst in term of performance, it takes forever to load and does not offer a functioning automation API. 

As for clearing false positives, WebStore have been the easiest among the three in my experience, simply replying the automatic email is enough to get a human review. Unlike the other two which require submitting a new version with an explanation letter.

**Comment by okiehsch** [comment]
Point taken, you are clearly in a better position to make an argument about the quality of CWS than I am.
I can only go by anecdotal evidence and reading through https://github.com/dessant/search-by-image/issues/57 especially the replies by google felt like a kafkaesque experience to me.

**Comment by jspenguin2017** [comment]
Every store have some weird policies. Microsoft Store doesn't allow screenshots to have description on the top 75% of the image, but it's perfectly fine on the bottom. AMO doesn't allow `$` to refer to a library that's not a trusted (hash whitelisted) version of jQuery, but it's perfectly fine after renaming the namespace. 

I do agree that WebStore emails have poorer quality than the other two stores and I do wish WebStore emails would contains proper description and actionable directives like those from the other two stores, but I don't think this is something to freak out about.
